add objective function to define feature visualization process

ResearchDaniel opened the pull request:
**Describe your improvements**
- Added layer-level deep dream visualization and refactored to promote code reusability. Different objectives can now be used to do filter/layer level visualization. 
- Separated image transformation so that it is only performed for loss computation, not for optimization. The standard transformations did not work as expected otherwise.

Using standard_transformation, before/after this PR (`images.initialize_image_ref(299, 299, fft=False, decorrelate=False, seed=1)`)
![inceptionv3_optimize_transformed](https://user-images.githubusercontent.com/8738140/177168369-22be365c-8430-427d-9785-b754ebb1c033.png)
![inceptionv3_optimize_param](https://user-images.githubusercontent.com/8738140/177168412-08e7053a-4a36-4fb3-9c6a-4a28d7832f6f.png)

Sparkier commented on luna/featurevis/objectives.py:
```diff
@@ -0,0 +1,86 @@
+import tensorflow as tf
+
+class FilterObjective(object):
```
I think we should be consistent between channel/filter.

Sparkier commented on docs/source/usage.rst:
```diff
@@ -31,7 +31,9 @@ The following provides an example of how Luna can be used for feature visualizat
 
     opt_param = featurevis.OptimizationParameters(iterations, learning_rate, optimizer=optimizer)
-    activation, image= featurevis.visualize(image, model, layer="mixed5", optimization_parameters=opt_param, filter_index=30, transformation=my_trans)
+    # Visualize channel 30 within the mixed5 layer
+    objective = featurevis.objectives.FilterObjective(model, layer="mixed5", filter_index=30)
```
channel vs. filter

Should be consistent.

Sparkier commented on luna/featurevis/featurevis.py:
```diff
@@ -22,28 +22,23 @@ class OptimizationParameters():
-    layer,
+    objective,
     optimization_parameters,
     filter_index=None,
```
Can't we remove this now?

ResearchDaniel replied:
Good catch

ResearchDaniel replied:
Ah, it was a merge conflict fix error

Sparkier commented on luna/featurevis/featurevis.py:
```diff
@@ -86,47 +80,78 @@ def visualize(
-def gradient_ascent_step(img, model, regularization, optimization_parameters,
-                         minimize, filter_index=None):
-    """Performing one step of gradient ascend.
+def gradient_ascent_step(img, objective, optimization_parameters,
+                         minimize):
```
Nit: can be one line now.

Sparkier commented on luna/featurevis/images.py:
```diff
@@ -124,11 +126,13 @@ def initialize_image_ref(
             image_f[..., :3], decorrelate=decorrelate, sigmoid=True)
+    else:
+        output = tf.nn.sigmoid(image_f)
+
+        
```
Blank lines should be removed.

Sparkier reviewed (commented):
Only minor changes from my side. I think this is the preferred implementation, mine was more of a hot fix.

I will let Fuzhan add her review, however. I think by now she knows the code best.

fuzhanrahmanian commented on luna/featurevis/objectives.py:
```diff
@@ -0,0 +1,86 @@
+        """
+        self.model = get_feature_extractor(model, layer)
+        self.regularization = regularization
+
```
Since pylint does not like classes with just one public method, it could be useful to implement a standard __repr__ function:

```suggestion
def __repr__(self) -> str:
        return f"LayerObjective({self.model}, {self.regularization})"
```

Alternatively, we could just deactivate the pylint error at the beginning of the class with:

" # pylint: disable=R0903 "

fuzhanrahmanian commented on luna/featurevis/objectives.py:
```diff
@@ -0,0 +1,86 @@
+    layer = model.get_layer(name=layer_name)
+    return tf.keras.Model(inputs=model.inputs, outputs=layer.output)
+
+
```
Extra white line. I think pylint also does not like this

fuzhanrahmanian commented on luna/featurevis/objectives.py:
```diff
@@ -0,0 +1,86 @@
+        self.model = get_feature_extractor(model, layer)
+        self.filter_index = filter_index
+        self.regularization = regularization
+
```
Since pylint does not like classes with just one public method, it could be useful to implement a standard __repr__ function:

```suggestion
def __repr__(self) -> str:
        return f"FilterObjective({self.model},\
                   {self.filter_index}, {self.regularization})"
```

Alternatively, we could just deactivate the pylint error at the beginning of the class with:

" # pylint: disable=R0903 "

fuzhanrahmanian reviewed (changes requested):
This is a very good way for distinguishing between Layer and Filter. Like the idea a lot to put it in separate classes. 
I run my local test and the functionality is intact. 

Please review some minor suggestion to make pylint and travis happy.

ResearchDaniel commented:
@fuzhanrahmanian Thanks for the very helpful comments. I installed pylint and fixed all your suggestions.

fuzhanrahmanian commented on luna/featurevis/images.py:
```diff
@@ -146,16 +148,6 @@ def fft_image(img, std=None, decay_power=1):
     Returns:
         New image in spatial domain.
     """
-    if tf.compat.v1.keras.backend.image_data_format() == "channels_first":
```
I think that this check is still necessary.
Since the function is accepting a shape that can be either channel first or not (see line 116 or 118), batch, channels, height, width as a general shape becomes incorrect. 

A "channel-last" model, would yield batch, channels, height, width = 1  299 3 3 which is not correct. 
A test with 
`images.initialize_image_ref(299, 299, fft=True, decorrelate=True, seed=1)`
would produce an error.

ResearchDaniel replied:
I had missed to push my the last commit that I think fixes this. Can you verify with the latest changes?

fuzhanrahmanian reviewed (changes requested):
Thank you, Daniel, for the changes. Nice that pylint helps you.
However, my test was not successful, especially regarding the shapes when calling FFT. I think that we should still separate between channel first and channel last when defining the shape.

fuzhanrahmanian reviewed (changes requested):
With this change, the color seems to disappear almost completely from the result. In mini-lucid, the gradient is calculated with activation and not loss (if you are trying to resemble mini-lucid that is).

I would suggest fixing the gradient override instead of completely removing it. 

Small reminder @ResearchDaniel: the previous issues about the shape of the last review are still open.

ResearchDaniel commented on luna/featurevis/images.py:
```diff
@@ -118,15 +115,13 @@ def initialize_image_ref(
-            size=[shape[0], shape[1],
-                  shape[2], shape[3]], scale=std
-        ).astype(np.float32)
+        shape = _to_img_shape(width, height, channels)
```
Fixed

ResearchDaniel commented:
> In mini-lucid, the gradient is calculated with activation and not loss (if you are trying to resemble mini-lucid that is)

I don't think that I changed the gradient calculation. I just renamed it

fuzhanrahmanian commented on luna/featurevis/featurevis.py:
```diff
@@ -153,27 +188,19 @@ def gradient_ascent_step(img, model, regularization, optimization_parameters,
-            grads_relu_0 = rg.redirected_relu_grad(img, grads*-1)
-            grads_modified = rg.redirected_relu6_grad(img, grads_relu_0)
-            optimization_parameters.optimizer.apply_gradients(zip([grads_modified], [img]))
+            # Relu gradient overrides do not seem to work,
```
I would suggest keeping the current version of the gradient override as by removing it, the problem still persists (at least with this gradient we are getting some minimal coloration)

ResearchDaniel replied:
These are the results I get with/without the current gradient overrides.
**With gradient overrides. Transformations: scale + blur**
![test_scale_blur_with_gradient_overrides](https://user-images.githubusercontent.com/8738140/182085974-75b88cfe-2243-464b-91c8-85f144edf0bd.png)
**Without gradient overrides. Transformations: scale + blur**
![test_scale_blur_without_grad_override](https://user-images.githubusercontent.com/8738140/182085976-582c7512-6861-4ebc-91fc-d81502e68e79.png)
**With gradient overrides. Transformations: default**
![test_default_trans_with_gradient_overrides](https://user-images.githubusercontent.com/8738140/182086686-0383a036-3ac9-49a7-aa42-ab114169761a.png)
**Without gradient overrides. Transformations: default**
![test_default_transformations_without_gradient_overrides](https://user-images.githubusercontent.com/8738140/182087097-e9d43ecd-ffea-4a68-b997-15f998b04753.png)


I think that separating the transformations out of the the optimized input is the reason why the gradient overrides now impact the result (more?).

Sparkier replied:
So what is our conclusion? We remove them for now and add them when we have an idea how to make them work?

fuzhanrahmanian replied:
Running the test file, at the moment, the result without gradient override looks like this:
![test](https://user-images.githubusercontent.com/52014627/182245458-bdd5fb77-14c1-45c0-b61b-6e57377a27df.png)
 
In my opinion, a "broken" gradient override is still a better result regarding both color and sharpness than something completely blurry but if you both would like to remove it, I am not against it.

ResearchDaniel commented on luna/featurevis/objectives.py:
```diff
@@ -0,0 +1,139 @@
+    def __repr__(self) -> str:
+        return f"LayerObjective({self.model}, {self.regularization})"
+
+class LayerActivationObjective:
```
@Sparkier

Sparkier commented on luna/featurevis/objectives.py:
```diff
@@ -0,0 +1,139 @@
+        Returns:
+            number: the loss for the specified setting
+        """
+
```
Nit: remove blank line.

Sparkier replied:
Nit: remove blank line.

Sparkier reviewed (approved):
Only thing remaining is the gradient override. I think we should probably remove it for now and make an issue that we address soon to look into it again.
